File: js/lib/UniversalId.ts
import { Gmod, Codebooks, LocalIdParsingErrorBuilder, LocalId } from ".";
import { Locations } from "./Location";
import { UniversalIdBuilder } from "./UniversalId.Builder";

export class UniversalId {
    private _builder: UniversalIdBuilder;
    private _localId: LocalId;

    public constructor(builder: UniversalIdBuilder) {
        if (!builder.isValid) throw new Error("Invalid UniversalId state");
        this._builder = builder;
        this._localId = builder.localId!.build();
    }

    public get localId(): LocalId {
        return this._localId;
    }

    public get builder(): UniversalIdBuilder {
        return this._builder;
    }

    public get imoNumber() {
        return this._builder.imoNumber!;
    }

    public equals(other?: UniversalId): boolean {
        if (!other) return false;
        return this._builder.equals(other.builder);
    }

    public toString() {
        return this._builder.toString();
    }

    public static parse(
        universalId: string,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return new UniversalId(
            UniversalIdBuilder.parse(
                universalId,
                gmod,
                codebooks,
                locations,
                errorBuilder
            )
        );
    }

    public static async parseAsync(
        universalIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return (
            await UniversalIdBuilder.parseAsync(universalIdString, errorBuilder)
        ).build();
    }
}

File: cpp/src/dnv/vista/sdk/UniversalId.cpp
#include "pch.h"

#include "dnv/vista/sdk/UniversalId.h"

#include "dnv/vista/sdk/ImoNumber.h"
#include "dnv/vista/sdk/LocalId.h"
#include "dnv/vista/sdk/ParsingErrors.h"
#include "dnv/vista/sdk/UniversalIdBuilder.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// UniversalId class
	//=====================================================================

	//----------------------------------------------
	// Construction / destruction
	//----------------------------------------------

	UniversalId::UniversalId( const UniversalIdBuilder& builder )
		: m_imoNumber( *builder.imoNumber() ),
		  m_localId( builder.localId()->build() )
	{
		if ( !builder.isValid() )
		{
			throw std::invalid_argument( "Invalid UniversalIdBuilder state" );
		}
	}

	//----------------------------------------------
	// String conversion
	//----------------------------------------------

	std::string UniversalId::toString() const
	{
		std::ostringstream oss;

		oss << UniversalIdBuilder::namingEntity;
		oss << "/" << m_imoNumber.toString();
		oss << m_localId.toString();

		return oss.str();
	}

	//----------------------------------------------
	// Static parsing methods
	//----------------------------------------------

	UniversalId UniversalId::parse( std::string_view universalIdStr )
	{
		auto builder = UniversalIdBuilder::parse( universalIdStr );

		return builder.build();
	}

	bool UniversalId::tryParse( std::string_view universalIdStr, ParsingErrors& errors, std::optional<UniversalId>& universalId )
	{
		std::optional<UniversalIdBuilder> builder;

		if ( !UniversalIdBuilder::tryParse( universalIdStr, errors, builder ) )
		{
			universalId = std::nullopt;
			return false;
		}

		universalId = builder->build();

		return true;
	}
}
